common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/elastic/ElasticRetriever.scala

```diff
@@ -1,24 +1,29 @@
-package uk.ac.wellcome.pipeline_storage
+package uk.ac.wellcome.pipeline_storage.elastic
 
-import scala.concurrent.{ExecutionContext, Future}
-import scala.util.{Failure, Success}
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
 
+import com.sksamuel.elastic4s.requests.get.{GetRequest, GetResponse}
 import com.sksamuel.elastic4s.{
   ElasticClient,
   Index,
   RequestFailure,
   RequestSuccess
 }
 import com.sksamuel.elastic4s.ElasticDsl._
-import com.sksamuel.elastic4s.circe._
-import io.circe.Decoder
 import grizzled.slf4j.Logging
+import uk.ac.wellcome.pipeline_storage.{
+  Retriever,
+  RetrieverMultiResult,
+  RetrieverNotFoundException
+}
+
+trait ElasticRetriever[T] extends Retriever[T] with Logging {
+  val client: ElasticClient
+  val index: Index
 
-class ElasticRetriever[T](client: ElasticClient, index: Index)(
-  implicit val ec: ExecutionContext,
-  decoder: Decoder[T])
-    extends Retriever[T]
-    with Logging {
+  def createGetRequest(id: String): GetRequest
+  def parseGetResponse(response: GetResponse): Try[T]
```

Comment:
Couldn't we provide default implementations for these so this remains a `class` instead of a `trait`? I am not a fan of lots of this sort of inheritance generally, but I think it would be appropriate here and avoid the boiler plate of having to create an `ElasticSourceRetriever` for cases which don't need custom behaviour

Reply:
@alexwlchan did you miss this comment?

Reply:
Ah sorry. I considered it, but I'm not a super fan of classes extending classes tbh. We use the `trait … class` pattern everywhere else, and the difference between SourceRetriever and Retriever is very slight.

Reply:
I have taken a look at the code again and I think I misread it the first time around, this is fine by me :+1: